src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -1,70 +1,113 @@
             IConsensusRules consensusRules,
-            IFinalizedBlockHeight finalizedBlockHeight)
+            IFinalizedBlockHeight finalizedBlockHeight,
+            IConnectionManager connectionManager,
+            IBlockStore blockStore = null)
         {
             this.network = network;
             this.chainState = chainState;
+            this.blockValidator = blockValidator;
             this.consensusSettings = consensusSettings;
-            this.concurrentChain = concurrentChain;
-            this.lookaheadBlockPuller = lookaheadBlockPuller;
+            this.blockPuller = blockPuller;
             this.consensusRules = consensusRules;
+            this.connectionManager = connectionManager;
+            this.blockStore = blockStore;
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
 
-            this.chainedHeaderTree = new ChainedHeaderTree(network, loggerFactory, chainedHeaderValidator, checkpoints, chainState, finalizedBlockHeight, consensusSettings);
+            this.chainedHeaderTree = new ChainedHeaderTree(network, loggerFactory, blockValidator, checkpoints, chainState, finalizedBlockHeight, consensusSettings);
+
+            this.treeLock = new object();
+            this.reorgLock = new object();
+            this.blockRequestedLock = new object();
 
-            this.blocksRequested = new Dictionary<uint256, List<Action>>();
-            this.toDownloadQueue = new AsyncQueue<(ChainedHeader headerFrom, ChainedHeader headerTo)>();
+            this.blocksRequested = new Dictionary<uint256, List<OnBlockDownloadedCallback>>();
+            this.toDownloadQueue = new Queue<BlockDownloadRequest>();
         }
 
-        public async Task InitializeAsync()
+        /// <summary>
+        /// Set the tip of <see cref="ConsensusManager"/> to be the header that is stored to disk,
+        /// if the given <see cref="consensusTip"/> is not in store then rewind store till a tip is found.
+        /// If <see cref="IChainState.BlockStoreTip"/> is not null (block store is enabled) ensure the store and consensus manager are on the same tip,
+        /// if consensus manager is ahead then reorg it to be the same height as store.
+        /// </summary>
+        /// <param name="consensusTip">The consensus tip that is attempted to be set.</param>
+        public async Task InitializeAsync(ChainedHeader consensusTip)
         {
-            this.logger.LogTrace("()");
+            this.logger.LogTrace("({0}:'{1}')", nameof(consensusTip), consensusTip);
+
+            // TODO: consensus store
+            // We should consider creating a consensus store class that will internally contain
+            //  coinview and it will abstract the methods `RewindAsync()` `GetBlockHashAsync()` 
 
             uint256 utxoHash = await this.consensusRules.GetBlockHashAsync().ConfigureAwait(false);
             bool blockStoreDisabled = this.chainState.BlockStoreTip == null;
 
             while (true)
```

Comment:
Some kind of limit on this loop should be defined – possibly one block store rewound to Genesis

Reply:
Maybe use the max reorg flag?

Reply:
How about bitcoin?